### ISSUE TITLE
SPServiceAppSecurity is not modifying members with "Members" parameter

### ISSUE THREAD

**Issue by wickedave** [PowerShell/SharePointDsc#641]
**Details of the scenario you try and problem that is occurring:**
I'm trying to set Administrators permissions on a service application with SPServiceAppSecurity resource.
When using "Members" parameter, it's detecting it is not matching but nothing was changed.

Here is the verbose when it is running:
```
[[SPServiceAppSecurity]SAPermission-UPA-01-Lab-SharingPermissions]
VERBOSE: [SP-Server]: LCM:  [ Start  Resource ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
VERBOSE: [SP-Server]: LCM:  [ Start  Test     ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Testing all security options for Administrators in Lab-UPA-01
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Getting all security options for Administrators in Lab-UPA-01
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Executing as the local run as user UBISOFT-ORG\svc_sps_setup
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Processing Members parameter
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Security list does not match
VERBOSE: [SP-Server]: LCM:  [ End    Test     ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
in 0.3440 seconds.
VERBOSE: [SP-Server]: LCM:  [ Start  Set      ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Setting all security options for Administrators in Lab-UPA-01
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Getting all security options for Administrators in Lab-UPA-01
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Executing as the local run as user UBISOFT-ORG\svc_sps_setup
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Executing as the local run as user domain\username
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplication.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Get-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Grant-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Grant-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Grant-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of New-SPClaimsPrincipal.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Revoke-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Revoke-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Revoke-SPObjectSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving BeginProcessing Method of Set-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving ProcessRecord Method of Set-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]:                            [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
Leaving EndProcessing Method of Set-SPServiceApplicationSecurity.
VERBOSE: [SP-Server]: LCM:  [ End    Set      ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
in 2.5620 seconds.
VERBOSE: [SP-Server]: LCM:  [ End    Resource ]  [[SPServiceAppSecurity]SAPermission-UPA-01-Lab-Administrators]
```

**The DSC configuration that is using the resource:**
```
Configuration Example 
{
    param(
        [Parameter(Mandatory = $true)]
        [PSCredential]
        $SetupAccount
    )
    Import-DscResource -ModuleName SharePointDsc

    node localhost {
        $membersToInclude = @()
        $membersToInclude += MSFT_SPServiceAppSecurityEntry {
                                Username    = "CONTOSO\SharePointFarmAccount"
                                AccessLevel = "Full Control"
                            }
        $membersToInclude += MSFT_SPServiceAppSecurityEntry {
                                Username    = "CONTOSO\SharePointServiceApps"
                                AccessLevel = "Full Control"
                            }
        SPServiceAppSecurity UserProfileServiceSecurity
        {
            ServiceAppName  = "User Profile Service Application"
            SecurityType         = "Administrators"
            Members              = $membersToInclude
            PsDscRunAsCredential = $SetupAccount 
        }
    }
}
```

**Version of the Operating System and PowerShell the DSC Target Node is running:**
Windows server 2016 with SharePoint 2016

**Version of the DSC module you're using:**
1.8.0.0

**Comment by lazda** [comment]
Additionally, access levels are wrong in MSFT_SPServiceAppSecurityEntry ( from .mof: {"Change Permissions","Full Control", "Write", "Read"} )
Access level depends on ServiceApplication type.

for Administrators schould be:
Create Target Application
Delete Target Application
Manage Target Application
All Target Applications
Manage Profiles
Manage Audiences
Manage Permissions
Retrieve People Data for Search Crawlers
Manage Social Data
Full Control
Read (Diagnostics Pages Only)

For SharingPermissions:
Read Access to Term Store
Read and Restricted Write Access to Term Store
Full Access to Term Store
Read Access to Term Store
Read and Restricted Write Access to Term Store
Full Access to Term Store
Full Control

and better is to replace Revoke-SPObjectSecurity when replacing access levels for principal (for Members & MembersToInclude). Use Grant-SPObjectSecurity with -Replace switch.
Reference https://docs.microsoft.com/en-us/powershell/module/sharepoint-server/grant-spobjectsecurity?view=sharepoint-ps

**Comment by ykuijs** [comment]
Fix will be included in my next Bugfix PR

**Comment by ykuijs** [comment]
Creating new issue for extending the access levels

**Comment by ykuijs** [comment]
Closed via PR #826
